### DISCUSSION TITLE
concatenate performance

### DISCUSSION THREAD

**Discussion by atravert** [spectrochempy/spectrochempy#408] [Developer discussions]
`concatenate` performs many checks which limits performance when hundreds or thousands datasets have to be concatenated. In particular the execution timpe does not scale well and increase nearly as N**2  (orange curve below, for large sets of spa files)), which makes impossible concatenation of thousands of datasets.  
In many cases, they are usually homogeneous in all dimensions except the one along which the concatenation has to be made. It could be useful to add a `homogeneous` flag in  `importer()` and `concatenate()` that would skip these tests. I have made it on this branch: [feature/concat_homo](https://github.com/atravert/spectrochempy/tree/feature/concat_homo), and got much faster execution and nearly linear scaling (blue curve):

![image](https://user-images.githubusercontent.com/10098314/157646933-dc2a556d-88a6-42e1-908a-7dc75b9a4f6d.png)

@fernandezc ,  as you are currently making a lot of refactoring, i don't know if it is worthwhile adding tests and making a PR. Let me know.

**Comment by fernandezc** [comment]
Yes indeed there are a lot of places where we can make improvements in performance. I don't know if it's the best solution to add more parameters. Is there any way to detect the fact that the arrays are homogeneous automatically? 

In any case, don't take into account my refactoring: it will take me a long time and the current program must evolve (for the better).

**Comment by atravert** [comment]
yez, good idea to avoid supplemen,tary keywords: a solution could be to try first concatenate without check, and catch errors, then if error concatenate with the current procedure  ?   
```
try:
   concatenate assuming homogeneous
   except Exception:
      try:
         concatenate with current checks (fix units differences, missing labels, etc...)
```

**Reply by fernandezc** [reply]
Let's give it a try! 👍

**Reply by atravert** [reply]
Thanks !
